pkg/skaffold/build/cache/cache.go

```diff
@@ -21,90 +21,81 @@ import (
 	homedir "github.com/mitchellh/go-homedir"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 	yaml "gopkg.in/yaml.v2"
 )
 
+// For testing
+var (
+	newDockerClient = createDockerClient
+)
+
+// ImageDetails holds the Digest and ID of an image
+type ImageDetails struct {
+	Digest string `yaml:"digest,omitempty"`
+	ID     string `yaml:"id,omitempty"`
+}
+
 // ArtifactCache is a map of [artifact dependencies hash : ImageDetails]
 type ArtifactCache map[string]ImageDetails
 
-// Cache holds any data necessary for accessing the cache
-type Cache struct {
+// cache holds any data necessary for accessing the cache
+type cache struct {
 	artifactCache      ArtifactCache
+	dependencies       DependencyLister
 	client             docker.LocalDaemon
-	builder            build.Builder
-	imageList          []types.ImageSummary
-	cacheFile          string
 	insecureRegistries map[string]bool
-	useCache           bool
-	isLocalBuilder     bool
-	pushImages         bool
-	localCluster       bool
-	prune              bool
+	cacheFile          string
+	imagesAreLocal     bool
 }
 
-var (
-	// For testing
-	localCluster    = config.GetLocalCluster
-	remoteDigest    = docker.RemoteDigest
-	newDockerClient = docker.NewAPIClient
-	noCache         = &Cache{}
-)
+// DependencyLister fetches a list of dependencies for an artifact
+type DependencyLister interface {
+	DependenciesForArtifact(ctx context.Context, artifact *latest.Artifact) ([]string, error)
+}
 
 // NewCache returns the current state of the cache
-func NewCache(builder build.Builder, runCtx *runcontext.RunContext) *Cache {
+func NewCache(runCtx *runcontext.RunContext, imagesAreLocal bool, dependencies DependencyLister) (Cache, error) {
 	if !runCtx.Opts.CacheArtifacts {
-		return noCache
+		return &noCache{}, nil
 	}
-	cf, err := resolveCacheFile(runCtx.Opts.CacheFile)
+
+	cacheFile, err := resolveCacheFile(runCtx.Opts.CacheFile)
 	if err != nil {
 		logrus.Warnf("Error resolving cache file, not using skaffold cache: %v", err)
-		return noCache
+		return &noCache{}, nil
 	}
-	cache, err := retrieveArtifactCache(cf)
+
+	artifactCache, err := retrieveArtifactCache(cacheFile)
 	if err != nil {
 		logrus.Warnf("Error retrieving artifact cache, not using skaffold cache: %v", err)
-		return noCache
-	}
-	client, err := newDockerClient(runCtx.Opts.Prune(), runCtx.InsecureRegistries)
-	if err != nil {
-		logrus.Warnf("Error retrieving local daemon client; local daemon will not be used as a cache: %v", err)
-	}
-	var imageList []types.ImageSummary
-	if client != nil {
-		imageList, err = client.ImageList(context.Background(), types.ImageListOptions{})
-		if err != nil {
-			logrus.Warn("Unable to get list of images from local docker daemon, won't be checked for cache.")
-		}
+		return &noCache{}, nil
 	}
 
-	lc, err := localCluster()
-	if err != nil {
-		logrus.Warn("Unable to determine if using a local cluster, cache may not work.")
+	client, err := newDockerClient(runCtx)
```

Comment:
should we even be trying to create this client if `imagesAreLocal == false`?

Reply:
yes for eg when images where built locally and now we switch to a remote cluster and just need to push them.